### ISSUE TITLE
w3c-test.org is being super slow

### ISSUE THREAD

**Issue by marcoscaceres** [web-platform-tests/wpt#7444]
Making request to https://w3c-test.org is being super slow, like 1 minute to respond... and sometimes just times out. 

https://w3c-test.org/payment-request/payment-request-abort-method.https.html

This obviously makes it hard to share tests with people who don't have WPT installed locally.

**Comment by marcoscaceres** [comment]
(sorry Mike, wasn't sure who to bug about this)

**Comment by marcoscaceres** [comment]
This has been happening for about 1-2 weeks.

**Comment by sideshowbarker** [comment]
> Making request to https://w3c-test.org is being super slow, like 1 minute to respond... and sometimes just times out.

Please try it now

> This has been happening for about 1-2 weeks.

Unfortunately, it’s going to continue to happen. We make zero guarantees about the reliability or availability of https://w3c-test.org/

> This obviously makes it hard to share tests with people who don't have WPT installed locally.

Yup, understood. But you have the option of running your own instance on the web somewhere, and using that to share tests

I wish I had a more-helpful answer here, but I don’t

**Comment by marcoscaceres** [comment]
> I wish I had a more-helpful answer here, but I don’t. 

That's ok... maybe we can throw server power at it? We could ask Mozilla or one of the other companies to foot the bill?

**Comment by sideshowbarker** [comment]
> That's ok... maybe we can throw server power at it? We could ask Mozilla or one of the other companies to foot the bill?

It’s just a problem with wptserve itself — the python process — not a problem with resources at all.

The machine I run https://w3c-test.org/ from has 8GB of RAM, ~6.5GB of which is completely unused at any given time, and 4 fast cores which `top` shows are only under very modest load at any time.

So the wptserve problems aren’t related to excessive load on the system itself nor to resource limitations in the underlying system.

Instead the problems are strictly in wptserve itself. It gets itself wedged in ways that I don’t understand. And I’m sure that anybody else running wptserve as persistent long-running process under any kind of load at all — regardless of the environment — is eventually going to run into exactly the same problems I observe with it in the https://w3c-test.org/ environment.

The only reason we don’t have confirmation of that is, there’s nobody else yet who has set up wptserve to run as a persistent process in the way we have on https://w3c-test.org/

**Comment by marcoscaceres** [comment]
Ok, got it. I guess I can just buy a new domain and serve just the payments tests from there if need be.  They don't need any fancy headers, redirects, etc.

**Comment by foolip** [comment]
@mdittmer has been monitoring w3c-test.org for a while now: https://bit.ly/ecosystem-infra-status

The uptime is really very bad, and I notice it once every few days myself.

@sideshowbarker, is there any easy fix for this situation, like throwing more hardware at it? Or would it take something more elaborate like 2+ machines take turns restarting the server frequently and a load balancer in front of them?

**Comment by sideshowbarker** [comment]
> @sideshowbarker, is there any easy fix for this situation

No. The fix is to make wptserve work correctly when it’s run persistently for long periods of time, and responding to webhooks for every PR that gets created or change.

As far as I know, w3c-test.org is the only environment where wptserve is run that way currently. I strongly suspect that anybody would have the same problems with it if they ran wptserve in the same way in some other environment.

But the only way to find out if that’s true would be for somebody to set up wptserve in another environment with it responding to requests for all the same webhooks that the w3c-test.org is.

So far nobody has taken the time to try that. Unless and until somebody does, the root causes of the problems are going to remain a mystery.

If somebody were to take to them to set up an instance elsewhere, I guess the best-case outcome would be, we discover that there aren’t actually any problems with wptserve but instead the problems we have been seeing are specific to w3c-test.org — and at that point I could retire from being the sysadmin for it and we could just shut it down and instead use the new (working) host from now on.

The next-best-case outcome would be that we observe the same problems with wptserve in the new system but because we have more people with access to it (or at least some people who know how to troubleshoot problem with long-running Python processes or heavy file IO or whatever), we finally get the problems figured out and fixed (in wptserve or wherever).

**Comment by foolip** [comment]
Oh, so we can't really look deeply into what's going wrong with w3c-test.org, then, because we're not totally in control of the machine?

Setting up another instance would presumably be pretty straightforward, do you have all of the config saved somewhere? But to get the same kind of load to see if the problem is the same wouldn't be trivial, although I guess one could just hammer it and see.

**Comment by sideshowbarker** [comment]
> Oh, so we can't really look deeply into what's going wrong with w3c-test.org

Sure we can

> because we're not totally in control of the machine?

I am totally in control of the machine and we can do whatever we want with it. If we have anybody who’s serious about taking time to investigate the problems, I can give them shell access to the server environment, and root, so they can go in and look at it all directly.

**Comment by foolip** [comment]
@gsnedders, can you summarize what you were saying yesterday about this?

In any event, this is now blocked on somewhat spending time on it.

**Comment by foolip** [comment]
@lukebjerring, this is the issue I mentioned today, that we could end up incidentally fixing if we need to run wpt serve for an arbitrary commit for testing that commit against an OS like Android.

**Comment by mdittmer** [comment]
I followed the link in the fist comment here. It was fast. @marcoscaceres and @foolip is this still an issue?

**Comment by foolip** [comment]
Possibly related is that PRs aren't always mirrored as they should be, see https://github.com/web-platform-tests/wpt/pull/11355#issuecomment-394689193.

**Comment by gsnedders** [comment]
That isn't being slow, that's somehow requests failing to load its vendored packages if you look at the tracebacks. I'd call that unrelated.

**Comment by mdittmer** [comment]
Ping @sideshowbarker and @foolip. This is labeled roadmap, and been open since January. Is it approaching the top of anyone's stack?

**Comment by lukebjerring** [comment]
Yes; @jugglinmike is working on this (indirectly).

**Comment by foolip** [comment]
The replacement is http://web-platform-tests.live, @jugglinmike is working out how to transition. Everyone can test it, but don't direct a lot of traffic at it.

**Comment by jugglinmike** [comment]
@mdittmer would you mind extending [the Status Cake service](https://uptime.statuscake.com/?TestID=kbUT7hiAxw) to also monitor the new domain? (I believe you set that up--let me know if I'm mistaken.) Although [we've configured some health checks through AWS directly](https://github.com/bocoup/infrastructure-web-platform/blob/42765778a1f16dc1bb9731c16ba244b5c695bbe2/terraform/projects/web-platform-tests-live/network.tf#L128-L138), also having data collected by the same service that monitors w3c-test.org will make future performance comparisons more realistic (not to mention easier--[we already have a script to scrape data from Status Cake](https://github.com/w3c/specberus/issues/758#issuecomment-400480167)).

**Comment by foolip** [comment]
The "Directory listing: web-platform-tests" in https://uptime.statuscake.com/?TestID=kbUT7hiAxw is actually web-platform-tests.live. @mdittmer, can you rename it?

**Comment by jugglinmike** [comment]
I have another request (there's a pun there somewhere, but I don't have it in me right now): to more closely match the data for w3c-test.org, could the rate be set to 5 minutes instead of 15?

**Comment by foolip** [comment]
https://uptime.statuscake.com/?TestID=kbUT7hiAxw now has 5m updates for all the things. (15m is the default, as you might guess.)

**Comment by jugglinmike** [comment]
Excellent. Thank you!

**Comment by mdittmer** [comment]
_friendly ping_. Linked dashboard suggests that there are still reliability issues here. @sideshowbarker are there plans to work on this during this quarter?

**Comment by sideshowbarker** [comment]
> _friendly ping_. Linked dashboard suggests that there are still reliability issues here. @sideshowbarker are there plans to work on this during this quarter?

Yes, after this week

**Comment by mdittmer** [comment]
_friendly ping_. Has this improved but the issue not updated, or did it end up falling off the radar last quarter?

**Comment by foolip** [comment]
Closing. Maybe we'll get new reports, but they might not be for the same reason. Also, web-platform-tests.live will probably be ready for action soon.
